src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,257 @@
+  extends mutable.AbstractBuffer[A]
+    with mutable.Buffer[A]
+    with generic.GenericTraversableTemplate[A, ArrayDeque]
+    with mutable.BufferLike[A, ArrayDeque[A]]
+    with mutable.IndexedSeq[A]
+    with mutable.IndexedSeqOptimized[A, ArrayDeque[A]]
+    with mutable.Builder[A, ArrayDeque[A]]
+    with Serializable {
+
+  private[this] var mask = 0   // modulus using bitmask since array.length is always power of 2
+  set(array, start, end)
+
+  def this(initialSize: Int = ArrayDeque.defaultInitialSize) = this(ArrayDeque.alloc(initialSize), 0, 0)
+
+  override def apply(idx: Int) = {
+    checkIndex(idx)
+    get(idx).asInstanceOf[A]
+  }
+
+  override def update(idx: Int, elem: A) = {
+    checkIndex(idx)
+    set(idx, elem.asInstanceOf[AnyRef])
+  }
+
+  override def +=(elem: A) = {
+    ensureCapacity()
+    array(end) = elem.asInstanceOf[AnyRef]
+    end = (end + 1) & mask
+    this
+  }
+
+  override def +=:(elem: A) = {
+    ensureCapacity()
+    start = (start - 1) & mask
+    array(start) = elem.asInstanceOf[AnyRef]
+    this
+  }
+
+  override def ++=:(xs: TraversableOnce[A]) =
+    xs.foldRight(this)((x, coll) => x +=: coll).asInstanceOf[this.type]
+
+  override def insertAll(idx: Int, elems: scala.collection.Traversable[A]) = {
+    checkIndex(idx)
+    val src = elems.toBuffer
+    /*val finalLength = src.length + this.length
+    // Either we resize right away or move prefix right or suffix left
+    if (2*finalLength >= array.length) {
+      val array2 = ArrayDeque.alloc(finalLength)
+      arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = idx)
+      Array.copy(src = src, srcPos = 0, dest = array2, destPos = idx, length = src.length)
+      arrayCopy(dest = array2, srcStart = idx, destStart = idx + src.length, maxItems = size)
+      set(array = array2, start = 0, end = finalLength)
+    } else {*/
+    // TODO: choose to move prefix right or suffix left
+    val suffix = drop(idx)
+    end = (start + idx) & mask
+    this ++= src ++= suffix
+    /*}*/
+  }
+
+  override def remove(idx: Int, count: Int): Unit = {
+    checkIndex(idx)
+    if (count <= 0) return
+    val removals = (size - idx) min count
+    // If we are removing more than half the elements, its cheaper to start over
+    // Else, either move the prefix right or the suffix left - whichever is shorter
+    if(2*removals >= size) {
+      val array2 = ArrayDeque.alloc(size - removals)
+      arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = idx)
+      arrayCopy(dest = array2, srcStart = idx + removals - 1, destStart = idx, maxItems = size)
+      set(array = array2, start = 0, end = size - removals)
+    } else if (size - idx <= idx + removals) {
+      (idx until size) foreach {i =>
+        val elem = if (i + removals < size) get(i + removals) else null
+        set(i, elem)
+      }
+      end = (end - removals) & mask
+    } else {
+      (0 until (idx + removals)).reverse foreach {i =>
```

Comment:
@gmethvin / @Ichoran : changed 2 while loops: 
https://github.com/scala/collection-strawman/pull/49/commits/4f94cdd5bc541700b9553fa960cdc87a7c98188a

I think there is still scope for improvement by using Java's array copy here (atleast for setting nulls) - the previous section is tricky since it is moving a slice of array to itself...